src/interface/viewport.c

```diff
@@ -1113,30 +1394,148 @@ int sub_98197C(
-		rotation
-	);
-	return 1;
+	RCT2_GLOBAL(0xF1AD28, uint32) = 0;
+	RCT2_GLOBAL(0xF1AD2C, uint32) = 0;
+
+	//Not a paint struct but something similar
+	paint_struct *ps = RCT2_GLOBAL(0xEE7888, paint_struct*);
+	if ((uint32) ps >= RCT2_GLOBAL(0xEE7880, uint32)) {
+		return false;
+	}
+
+	ps->image_id = image_id;
+
+	uint32 image_element = image_id & 0x7FFFF;
+	rct_g1_element *g1Element = gfx_get_g1_element(image_element);
+
+	rct_xyz16 coord_3d = {
+		.x = x_offset, // ax
+		.y = y_offset, // cx
+		.z = z_offset
+	};
+
+	switch (rotation) {
+		case 1:
+			coord_3d.x = -y_offset;
+			coord_3d.y = x_offset;
+			break;
+
+		case 2:
+			coord_3d.x = -x_offset;
+			coord_3d.y = -y_offset;
+			break;
+
+		case 3:
+			coord_3d.x = y_offset;
+			coord_3d.y = -x_offset;
+			break;
+	}
+
+	coord_3d.x += RCT2_GLOBAL(0x9DE568, sint16);
+	coord_3d.y += RCT2_GLOBAL(0x9DE56C, sint16);
+
+	rct_xy16 map = coordinate_3d_to_2d(&coord_3d, rotation);
+
+	ps->x = map.x;
+	ps->y = map.y;
+
+	int left = map.x + g1Element->x_offset;
+	int bottom = map.y + g1Element->y_offset;
+
+	int right = left + g1Element->width;
+	int top = bottom + g1Element->height;
+
+	RCT2_GLOBAL(0xF1AD1C, uint16) = left;
+	RCT2_GLOBAL(0xF1AD1E, uint16) = bottom;
+
+	rct_drawpixelinfo *dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+
+	if (right <= dpi->x) return false;
+	if (top <= dpi->y) return false;
+	if (left >= (dpi->x + dpi->width)) return false;
+	if (bottom >= (dpi->y + dpi->height)) return false;
+
+	rct_xyz16 boundBoxOffset = {
+		.x = bound_box_offset_x,
+		.y = bound_box_offset_y,
+		.z = bound_box_offset_z,
+	};
+
+	rct_xyz16 boundBox = {
+		.x = bound_box_length_x,
+		.y = bound_box_length_y,
+		.z = bound_box_length_z
+	};
+
+	sint16 temp;
+	switch (rotation) {
+		case 0:
+			boundBox.x--;
+			boundBox.y--;
+			break;
+
+		case 1:
+			boundBoxOffset.x = bound_box_offset_y;
+			boundBoxOffset.y = -bound_box_offset_x;
+
+			boundBox.x--;
+			temp = boundBox.x;
+			boundBox.x = boundBox.y;
+			boundBox.y = temp;
+			boundBox.x = -boundBox.x;
+			break;
+
+		case 2:
+			boundBoxOffset.x = -boundBoxOffset.x;
+			boundBoxOffset.y = -boundBoxOffset.y;
+
+			boundBox.x = -boundBox.x;
+			boundBox.y = -boundBox.y;
+			break;
+
+		case 3:
+			boundBoxOffset.x = bound_box_offset_y;
+			boundBoxOffset.y = -bound_box_offset_x;
+
+			boundBox.y--;
+			temp = boundBox.x;
+			boundBox.x = boundBox.y;
+			boundBox.y = temp;
+			boundBox.y = -boundBox.y;
+			break;
+	}
+
+	ps->other_x = boundBox.x + boundBoxOffset.x + RCT2_GLOBAL(0x009DE568, sint16);
+	ps->other_y = boundBox.y + boundBoxOffset.y + RCT2_GLOBAL(0x009DE56C, sint16);
+
+	// TODO: check whether this is right. edx is ((bound_box_length_z + bound_box_offset_z) << 16 | bound_box_offset_z)
+	ps->some_x = boundBoxOffset.z;
+	ps->some_y = boundBoxOffset.z + boundBox.z;
+
+	ps->var_1A = 0;
+	ps->attached_x = boundBoxOffset.x + RCT2_GLOBAL(0x009DE568, sint16);
+	ps->attached_y = boundBoxOffset.y + RCT2_GLOBAL(0x009DE56C, sint16);
```

Comment:
It looks like other_x/y is bound box x1, y1 and attached_x/y is bound box x2, y2. Some_x/y is bound box z1,z2. Can you rename all of these vars.